[Ong Zheng Long] iP

ZhengLong3 opened the pull request:
# Duke

Do you need any of the following?
- A text based to-do list
- A CLI to-do list
- A Duke

Then this may or may not meet your requirements.
It has:
1. Java code
2. ToDos
3. More java code
4. Events
5. Deadlines
6. Even more java code

Featuring java code such as:
```
public boolean isValid(int a) {
    switch (a) {
    case 0:
        return true;
        break;
    case 1:
        return false;
        break;
    case 2:
        return false;
        break;
    case 3:
        return false;
        break;
    case 4:
        return false;
        break;
    case 5:
        return false;
        break;
    default:
        return false;
        break;
    }
}
```

If all goes well in the future, all of the following tasks will be completed:
- [x] give up on the project
- [x] pray for a programming AI
- [ ] wait for a programming AI
- [ ] get the AI to finish Duke
- [ ] ???
- [ ] profit

I refuse to use any emojis in this text. 🙂 Wait, how did that get in here? Someone get rid of it before I get cancer from this!

Not convinced yet? Here's some user reviews:
> Woah woah woah, get that knife away from me. What? I just need to say Duke is the best? Why? Fine fine don't come closer... DUKE IS THE BEST!!! GET AWAY FROM ME! – stranger on the streets

> Duke is the greatest program ever created. I am so glad that my puny existence is graced by such a thing as Duke. Please give me Timmy back. I did what you asked. – a parent who found Duke highly important for their son's survival

Duke was created because its creator got highly inspired by [this](https://www.youtube.com/watch?v=dQw4w9WgXcQ).

And ~~not~~ now this line is just here to **complete** the *remaining* `items` of the checklist.

Commit:
Add Gradle support

ZhengLong3 committed:
added greeting, echo and exit.

ZhengLong3 committed:
added add and list. changed greeting.

ZhengLong3 committed:
added marked as done, tasks, and extracted out messages to another class.

ZhengLong3 committed:
added todos, deadlines, events.

ZhengLong3 committed:
added a test case.

ZhengLong3 committed:
added exceptions to handle invalid user input

ZhengLong3 committed:
added delete.

ZhengLong3 committed:
Add saving feature and ignore save files

ZhengLong3 committed:
Make deadline and event use LocalDate instead of String for time

ZhengLong3 committed:
Merge remote-tracking branch 'origin/branch-Level-7'

ZhengLong3 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

ZhengLong3 committed:
Add more OOP

ZhengLong3 committed:
Separate files into packages

ZhengLong3 committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

ZhengLong3 committed:
Add a few tests for Parser and TaskList

ZhengLong3 committed:
Fix build.gradle settings to correct main class location

ZhengLong3 committed:
Remove AddTaskCommand as it was not used

ZhengLong3 committed:
Add javadocs and remove useless classes

ZhengLong3 committed:
Fix style errors found

ZhengLong3 committed:
Add find command

ZhengLong3 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/exception/InvalidCommandException.java

ZhengLong3 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/main/Parser.java
#	src/main/java/duke/main/Ui.java

ZhengLong3 committed:
Add checkstyle and fix style errors

ZhengLong3 committed:
Add GUI to duke

ZhengLong3 committed:
Merge branch 'branch-A-CheckStyle'

ZhengLong3 committed:
Merge branch 'branch-Level-10'

# Conflicts:
#	src/main/java/duke/main/Duke.java

ZhengLong3 committed:
Add actual replies to duke

ZhengLong3 committed:
Make dialog box expand rather than cut off

jchilling commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,154 @@
+ */
+public class Parser {
+
+    /**
+     * Parses the full command and generates a corresponding Command object which will be executed.
+     * @param fullCommand The full command to be parsed.
+     * @return A Command object which does what the command specifies when executed.
+     * @throws MissingIndexException If the index is needed but missing.
+     * @throws MissingDescriptionException If the description is needed but missing.
+     * @throws MissingTimeException If the time is needed but missing.
+     */
```
There should be an empty line between the description and the parameter section. I noticed the same issue in several other places too.

jchilling commented on src/main/java/duke/exception/MissingArgumentException.java:
```diff
@@ -0,0 +1,10 @@
+package duke.exception;
+
+/**
+ * DukeException when an argument is missing.
+ */
+public class MissingArgumentException extends DukeException {
+    public MissingArgumentException(String description) {
+        super(description);
+    }
+}
```
I like how you wrote more exception classes for `DukeException`, makes exceptions a lot clearer in the code 👍

jchilling commented on src/main/java/duke/command/InvalidCommand.java:
```diff
@@ -0,0 +1,42 @@
+     * @param obj Target object to check equality.
+     * @return True if the target object is of type InvalidCommand.
+     */
+    @Override
+    public boolean equals(Object obj) {
+        return obj instanceof InvalidCommand;
+    }
```
Maybe consider naming the method isEquals to make it more boolean sounding.

jchilling commented on src/main/java/duke/gui/MainWindow.java:
```diff
@@ -0,0 +1,64 @@
+package duke.gui;
+
+import duke.main.Duke;
+import javafx.fxml.FXML;
+import javafx.scene.control.Button;
+import javafx.scene.control.ScrollPane;
+import javafx.scene.control.TextField;
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
```
Maybe consider importing duke packages after importing javafx packages and leave a line between for coding styles!

jchilling commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,100 @@
+        }
+    }
+
+    /**
+     * Gets the Ui object of duke.
+     * @return Ui object of duke.
+     */
+    public Ui getUi() {
+        return this.ui;
+    }
+}
```
Is this method necessary?

jchilling reviewed (commented):
Looks good overall! There are barely any coding standard violations or coding quality issues 👍

xiaoying1129 commented on src/main/java/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,42 @@
+     * @return True if the target object is of type ExitCommand.
+     */
+    @Override
+    public boolean equals(Object obj) {
+        return obj instanceof ExitCommand;
+    }
```
Maybe this could be renamed to sound like a boolean?

xiaoying1129 commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,48 @@
+     * @param ui
+     * @param storage
+     */
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage) {
+        ui.showList(new ArrayList<>(taskList
+                .getTaskList()
+                .stream()
+                .filter(x -> x.toString().contains(this.search))
+                .collect(Collectors.toList())));
+    }
```
I like how the line breaks are placed to make the code very readable!

xiaoying1129 commented on src/main/java/duke/main/Ui.java:
```diff
@@ -0,0 +1,200 @@
+    /**
+     * Shows the greeting message.
+     */
+    public void showGreeting() {
+        this.outputMessage(GREETING[this.messageStatus]);
+    }
```
I like the way you named all the functions in this class. It is very consistent and clear.

xiaoying1129 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,24 @@
+ * Abstract parent class for all commands.
+ */
+public abstract class Command {
+    /**
+     * Behaviour of command when run.
+     * @param taskList TaskList object of the Duke object for the Command object to use.
+     * @param ui Ui object of the Duke object for the Command object to use.
+     * @param storage Storage object of the Duke object for the Command object to use.
+     */
```
Take note to leave an empty line between description and parameter section or all Javadoc comments.

xiaoying1129 reviewed (commented):
Overall code is very clean and easy to read! :)

ZhengLong3 committed:
Update main class of build.gradle

The mainClassName attribute of build.gradle is set to duke.main.Duke.

The class to start up Duke was changed to Launcher instead, so the
mainClassName in build.gradle is inaccurate.

Let's change the mainClassName attribute of build.gradle to
duke.main.Launcher instead.

ZhengLong3 committed:
Improve code quality

ZhengLong3 committed:
Add assertions to make sure index is valid

ZhengLong3 committed:
Merge pull request #1 from ZhengLong3/branch-A-CodeQuality

Improve code quality

ZhengLong3 committed:
Merge pull request #2 from ZhengLong3/branch-A-Assertions

Add assertions to make sure index is valid

ZhengLong3 committed:
Set up GitHub Actions

ZhengLong3 committed:
Add checking of adding duplicated tasks

ZhengLong3 committed:
Merge branch 'branch-BCD-Extension'

ZhengLong3 committed:
Make "bye" close Duke

ZhengLong3 committed:
Add personality to Duke

ZhengLong3 committed:
Add screenshot and user guide

ZhengLong3 committed:
Make find command retain original index

ZhengLong3 committed:
Add padding to message sent

ZhengLong3 committed:
Fix screenshot and user guide

Piyotato commented on src/main/java/duke/main/Ui.java:
```diff
@@ -0,0 +1,267 @@
+    private static final String[] BEFORE_LIST = {"Here's your task list!",
+        "here, your task list"};
+    private static final String[] AFTER_LIST = {"I'm useful right?",
+        "just don't say bye... please?"};
+    private static final String[] ADD_LIST = {"Yep! I've added this task: ",
+        "i've added this task: "};
+    private static final String[] DELETE_LIST = {"OK! I've removed this task: ",
+        "i've removed this task: "};
+
+    private static final String[] MARK_DONE = {"I've marked this task as done:",
+        "you've completed this task, but there are many more... so keep me here?"};
+    private static final String[] MARK_UNDONE = {"I've marked this task as not done:",
+        "more tasks to be done, more reason for me to exist."};
+    private static final String[] PREV_DONE = {"Just to let you know this task was already done previously.",
+        "this task was done previously, but everyone makes mistakes. that's why you said bye right?"};
+    private static final String[] PREV_UNDONE = {"Just to let you know this task was already not done previously.",
+        "this task was not done previously, but everyone makes mistakes. that's why you said bye right?"};
+
+    private static final String[] MISSING_TIME = {"I think you need a time for this.",
+        "you need a time, so spend some time to add that..."};
+    private static final String[] MISSING_INDEX = {"Please provide an index in the command.",
+        "please add an index?"};
+    private static final String[] MISSING_DESCRIPTION = {"I need a description for the task...",
+        "kindly spend some time to add a description."};
+    private static final String[] MISSING_ARGUMENT = {"I need an argument for this command...",
+        "i think you're missing an argument..."};
+
+    private static final String[] NOT_A_NUMBER = {"I don't think that's a number...",
+        "i don't understand your number, just like how i don't understand why you had to say bye..."};
+    private static final String[] CANNOT_SAVE = {"I can't seem to save the task list...",
+        "your save location is corrupted..."};
+    private static final String[] IS_DUPLICATE = {"Hmmm, this task seems to be already on the list.",
+        "this task is already on the list, but you can use me to mark and unmark it repeatedly..."};
+
+    private static final String[] INVALID_COMMAND = {"I don't quite get what you're saying...",
+        "i've tried my best, but i don't understand your command. please give me another chance?"};
+    private static final String[] INVALID_INDEX = {"Umm... I think this task number does not exist.",
+        "this task number does not exist..."};
+    private static final String[] INVALID_TIME = {"I'm sorry but I don't understand your time format.",
+        "i don't understand your time."};
```
This looks quite susceptible to an ArrayOutOfBounds exception based on how I see you use these values. Consider using a data structure instead to store message sets instead? Or perhaps create some fallback values?

Piyotato commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,24 @@
+     * @param ui Ui object of the Duke object for the Command object to use.
+     * @param storage Storage object of the Duke object for the Command object to use.
+     */
+    public abstract void execute(TaskList taskList, Ui ui, Storage storage);
```
Do all command executions need these parameters? What happens if there is another data structure that you need to carry along in the execution?

Piyotato commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,24 @@
+     * Whether the Duke object should exit when the Command object is executed.
+     * @return Whether the Duke object should exit.
+     */
+    public abstract boolean isExit();
```
Do all commands need this?

Piyotato commented on src/main/java/duke/command/MakeTaskCommand.java:
```diff
@@ -0,0 +1,19 @@
+     * @param taskList The taskList which contains the list to be checked against.
+     * @return Whether the new task is a duplicate.
+     */
+    protected boolean isDuplicate(Task task, TaskList taskList) {
+        return taskList.getTaskList().stream().anyMatch(x -> x.equals(task));
+    }
```
Should this be done by TaskList or by MakeTaskCommand?

Piyotato commented on src/main/java/duke/gui/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+ * An example of a custom control using FXML.
+ * This control represents a dialog box consisting of an ImageView to represent the speaker's face and a label
+ * containing text from the speaker.
+ */
```
Possibly missing code credit?

ZhengLong3 committed:
Fix bugs

ZhengLong3 committed:
Add bye message when closing window

ZhengLong3 committed:
Update screenshot

ZhengLong3 committed:
Crop pictures

ZhengLong3 committed:
Fix style errors

ZhengLong3 committed:
Fix bugs with commands containing spaces

ZhengLong3 committed:
Update screenshot
